lib/sandbox/sandbox_linux.go

```diff
@@ -0,0 +1,159 @@
+// +build linux
+
+package sandbox
+
+import (
+	"crypto/rand"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/docker/docker/pkg/mount"
+	"github.com/docker/docker/pkg/symlink"
+	"golang.org/x/sys/unix"
+)
+
+func isNSorErr(nspath string) error {
+	return ns.IsNSorErr(nspath)
+}
+
+func newNetNs() (*NetNs, error) {
+	netNS, err := ns.NewNS()
+	if err != nil {
+		return nil, err
+	}
+
+	return &NetNs{netNS: netNS, closed: false}, nil
+}
+
+func getNetNs(path string) (*NetNs, error) {
+	netNS, err := ns.GetNS(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NetNs{netNS: netNS, closed: false, restored: true}, nil
+}
+
+// NetNs handles data pertaining a network namespace
+type NetNs struct {
+	sync.Mutex
+	netNS    ns.NetNS
+	symlink  *os.File
+	closed   bool
+	restored bool
+}
+
+func (netns *NetNs) symlinkCreate(name string) error {
+	b := make([]byte, 4)
+	_, randErr := rand.Reader.Read(b)
+	if randErr != nil {
+		return randErr
+	}
+
+	nsName := fmt.Sprintf("%s-%x", name, b)
+	symlinkPath := filepath.Join(NsRunDir, nsName)
+
+	if err := os.Symlink(netns.netNS.Path(), symlinkPath); err != nil {
+		return err
+	}
+
+	fd, err := os.Open(symlinkPath)
+	if err != nil {
+		if removeErr := os.RemoveAll(symlinkPath); removeErr != nil {
+			return removeErr
+		}
+
+		return err
+	}
+
+	netns.symlink = fd
+
+	return nil
+}
+
+// Path returns the path of the network namespace handle
+func (netns *NetNs) Path() string {
+	if netns == nil || netns.netNS == nil { // TODO(vbatts) i'm not sure this is right. Why would this is netns.netNS be nil?
```

Comment:
Lets remove first check.  There is no way that "netns == nil", That should cause go to crash.

Reply:
updated to just
```golang
if netns.netNS == nil {
```
and the integration tests fail with nil pointer
```
# time="2018-05-03 14:37:12.695674452Z" level=debug msg="Calling TarUntar(/tmp/tmp.BjykOTXPas/crio/vfs/dir/19bdb9ec4283a34535efe1686a0387d79eecb4bd9326207128da21613c844006, /tmp/tmp.BjykOTXPas/crio/vfs/dir/940bddfaf2ebadf9d91ba7100ac3b1882a6d231aeb60e70911c4ca443fca559f)"
# time="2018-05-03 14:37:12.695707589Z" level=debug msg="TarUntar(/tmp/tmp.BjykOTXPas/crio/vfs/dir/19bdb9ec4283a34535efe1686a0387d79eecb4bd9326207128da21613c844006 /tmp/tmp.BjykOTXPas/crio/vfs/dir/940bddfaf2ebadf9d91ba7100ac3b1882a6d231aeb60e70911c4ca443fca559f)"
# time="2018-05-03 14:37:12.936462549Z" level=debug msg="created pod sandbox "a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64""
# time="2018-05-03 14:37:13.053248271Z" level=debug msg="pod sandbox "a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64" has work directory "/tmp/tmp.BjykOTXPas/crio/vfs-containers/a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64/userdata""
# time="2018-05-03 14:37:13.053388688Z" level=debug msg="pod sandbox "a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64" has run directory "/tmp/tmp.BjykOTXPas/crio-run/vfs-containers/a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64/userdata""
# time="2018-05-03 14:37:13.228141054Z" level=debug msg="mounted container "a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64" at "/tmp/tmp.BjykOTXPas/crio/vfs/dir/940bddfaf2ebadf9d91ba7100ac3b1882a6d231aeb60e70911c4ca443fca559f""  
# panic: runtime error: invalid memory address or nil pointer dereference
# [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xee9b32]
#
# goroutine 164 [running]:
# github.com/kubernetes-incubator/cri-o/lib/sandbox.(*NetNs).Path(0x0, 0xc42065f4d0, 0xc4205ed210)
#       /go/src/github.com/kubernetes-incubator/cri-o/lib/sandbox/sandbox_linux.go:80 +0x22
# github.com/kubernetes-incubator/cri-o/server.(*Server).RunPodSandbox(0xc42007c700, 0x7eff23890078, 0xc42027ce40, 0xc42062e0a0, 0x0, 0x0, 0x0)
#       /go/src/github.com/kubernetes-incubator/cri-o/server/sandbox_run.go:519 +0x3b0b
# github.com/kubernetes-incubator/cri-o/vendor/k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2._RuntimeService_RunPodSandbox_Handler(0x18e3ea0, 0xc42007c700, 0x7eff23890078, 0xc42027ce40, 0xc420062dc0, 0x0, 0x0, 0x0, 0x26141e0, 0xc4200140e0)
#       /go/src/github.com/kubernetes-incubator/cri-o/vendor/k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2/api.pb.go:4096 +0x290
# github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4203f2c80, 0x263a2c0, 0xc4203d7760, 0xc420079e00, 0xc42041ec00, 0x2608598, 0x0, 0x0, 0x0)
#       /go/src/github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc/server.go:843 +0xc41
# github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4203f2c80, 0x263a2c0, 0xc4203d7760, 0xc420079e00, 0x0)
#       /go/src/github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc/server.go:1040 +0x15a6
# github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420441290, 0xc4203f2c80, 0x263a2c0, 0xc4203d7760, 0xc420079e00)
#       /go/src/github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc/server.go:589 +0xa9
# created by github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
#       /go/src/github.com/kubernetes-incubator/cri-o/vendor/google.golang.org/grpc/server.go:590 +0xa1
# 2018/05/03 14:37:13 transport: http2Client.notifyError got notified that the client transport was broken EOF.
# time="2018-05-03T14:37:13Z" level=fatal msg="run pod sandbox failed: rpc error: code = Internal desc = transport is closing"
# 2018/05/03 14:37:13 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.BjykOTXPas/crio.sock: connect: connection refused"; Reconnecting to {/tmp/tmp.BjykOTXPas/crio.sock <nil>}
# time="2018-05-03T14:37:13Z" level=fatal msg="listing containers failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
# 2018/05/03 14:37:13 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.BjykOTXPas/crio.sock: connect: connection refused"; Reconnecting to {/tmp/tmp.BjykOTXPas/crio.sock <nil>}
# time="2018-05-03T14:37:13Z" level=fatal msg="listing pod sandboxes failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
# 0
# rm: cannot remove '/tmp/tmp.BjykOTXPas/crio-run/vfs-containers/a9ccff27c25ae2d2d034284d311fc892633805498e1a8972c3b39338e5864c64/userdata/shm': Device or resource busy
```

Reply:
tricky stuff. According to `./lib/sandbox/sandbox.go`:
```golang
331 // NetNs retrieves the network namespace of the sandbox
332 // If the sandbox uses the host namespace, nil is returned
333 func (s *Sandbox) NetNs() *NetNs {
334   return s.netns
335 }
```
So calling `sb.NetNs().Path()` is completely valid and would require `if netns == nil || netns.netNS == nil {`, though this feels nasty and is confusing from a readability perspective.

Reply:
Ok I will let it go.  But it feels nasty.